app/src/main/java/com/fivegmag/a5gmsmediasessionhandler/MediaSessionHandlerMessengerService.kt

```diff
@@ -302,5 +328,128 @@ class MediaSessionHandlerMessengerService() : Service() {
+    procedure when the generated random number is of a lower value than the samplePercentage value.
+    The Service Access Information indicating whether Consumption Reporting is provisioned for downlink streaming sessions is described in clause 11.2.3. When the clientConsumptionReportingConfiguration.samplePercentage value is 100, the Media Session Handler shall activate the consumption reporting procedure. If the samplePercentage is less than 100, the Media Session Handler shall generate a random number which is uniformly distributed in the range of 0 to 100, and the Media Session Handler shall activate the consumption report procedure when the generated random number is of a lower value than the samplePercentage value.
+     */
+    private fun IsConsumptionReportingActivated(): Boolean {
+        //if(currentServiceAccessInformation.isInitialized) {
+        if(currentServiceAccessInformation == null) {
+            return false
+        }
+
+        var samplePercentage: UInt =  currentServiceAccessInformation!!.clientConsumptionReportingConfiguration.samplePercentage;
+        if(samplePercentage > SamplePercentage_100.toUInt())
+        {
+            Log.i(TAG, "shilin>>>Invaild samplePercentage $samplePercentage in currentServiceAccessInformation")
+            return false;
+        }
+
+        if(SamplePercentage_100.toUInt() == samplePercentage)
+        {
+            return true;
+        }
+
+        val myRandomValue = Random.nextInt(0, SamplePercentage_100)
+        if(myRandomValue.toUInt() >= samplePercentage)
+        {
+            Log.i(TAG, "shilin>>>IsConsumptionReportingActivated true:myRandomValue[$myRandomValue],samplePercentage[$samplePercentage]")
+            return true;
+        }
+        else
+        {
+            Log.i(TAG, "shilin>>>IsConsumptionReportingActivated false:myRandomValue[$myRandomValue],samplePercentage[$samplePercentage]")
+            return false;
+        }
+    }
+
+    /* Refer to TS26.512 Clause 4.7.4
+                TS26.501 Clause 5.6.3*/
+    private fun NeedReportConsumption(): Boolean {
+        // check IsConsumptionReportingActivated
+        if (!IsConsumptionReportingActivated())
+        {
+            Log.i(TAG, "shilin>>>ReportConsumption not Activated")
+            return false;
+        }
+
+        // Start/stop of consumption of a downlink streaming session
+        // triggered when PlayerStates.PLAYING  &  PlayerStates.ENDED
+        /*if( Start || stop)
+        {
+            return true;
+        }*/
+
+        // check clientConsumptionReportingConfiguration.reportingInterval, start a timer
+        // reportConsumptionTimer()
+
+        // check clientConsumptionReportingConfiguration.locationReporting
+        // check clientConsumptionReportingConfiguration.accessReporting
+        if(currentServiceAccessInformation!!.clientConsumptionReportingConfiguration.locationReporting
```

Comment:
Why is consumption reporting only send if `locationReporting` or `accessReporting` is true?

Reply:
Just follow clause 4.7.4, as below:
If the consumption reporting procedure is activated, the Media Session Handler shall submit a consumption report to the 5GMSd AF when any of the following conditions occur:
-	Start of consumption of a downlink streaming session;
-	Stop of consumption of a downlink streaming session;
-	Upon determining the need to report ongoing 5GMS consumption at periodic intervals determined by the clientConsumptionReportingConfiguration.reportingInterval property.
_**-	Upon determining a location change, if the clientConsumptionReportingConfiguration.locationReporting property is set to True.**
**-	Upon determining an access network change (e.g. unicast to eMBMS, or vice versa), if the clientConsumptionReportingConfiguration.accessReporting property is set to True.**_

Did I misunderstand it?